tests/loggen/loggen.c

```diff
@@ -21,956 +20,400 @@
+  { "permanent", 'T', 0, G_OPTION_ARG_NONE, &global_plugin_option.permanent, "Send logs without time limit", NULL},
+  { "syslog-proto", 'P', 0, G_OPTION_ARG_NONE, &syslog_proto, "Use the new syslog-protocol message format (see also framing)", NULL },
+  { "sdata", 'p', 0, G_OPTION_ARG_STRING, &sdata_value, "Send the given sdata (e.g. \"[test name=\\\"value\\\"]\") in case of syslog-proto", NULL },
+  { "no-framing", 'F', G_OPTION_ARG_NONE, G_OPTION_ARG_NONE, &noframing, "Don't use syslog-protocol style framing, even if syslog-proto is set", NULL },
+  { "active-connections", 0, 0, G_OPTION_ARG_INT, &global_plugin_option.active_connections, "Number of active connections to the server (default = 1)", "<number>" },
+  { "idle-connections", 0, 0, G_OPTION_ARG_INT, &global_plugin_option.idle_connections, "Number of inactive connections to the server (default = 0)", "<number>" },
+  { "ipv6",    '6', 0, G_OPTION_ARG_NONE, &global_plugin_option.use_ipv6, "Use AF_INET6 sockets instead of AF_INET (can use both IPv4 & IPv6)", NULL },
+  { "csv", 'C', 0, G_OPTION_ARG_NONE, &csv, "Produce CSV output", NULL },
+  { "number", 'n', 0, G_OPTION_ARG_INT, &global_plugin_option.number_of_messages, "Number of messages to generate", "<number>" },
+  { "quiet", 'Q', 0, G_OPTION_ARG_NONE, &quiet, "Don't print the msg/sec data", NULL },
+  { "debug", 0, 0, G_OPTION_ARG_NONE, &debug, "Enable loggen debug messages", NULL },
+//  { "version",   'V', 0, G_OPTION_ARG_CALLBACK, arg_setup_callback, "Display version number (" SYSLOG_NG_PACKAGE_NAME " " SYSLOG_NG_VERSION ")", NULL },
+  { NULL }
+};
 
-static void
-time_val_diff_in_timeval(struct timeval *res, const struct timeval *t1, const struct timeval *t2)
+int
+get_debug_level(void)
 {
-  res->tv_sec = (t1->tv_sec - t2->tv_sec);
-  res->tv_usec = (t1->tv_usec - t2->tv_usec);
-  if (res->tv_usec < 0)
-    {
-      res->tv_sec--;
-      res->tv_usec += USEC_PER_SEC;
-    }
+  return debug;
 }
 
-static void
-time_val_add_time_val(struct timeval *res, const struct timeval *t1, const struct timeval *t2)
+/* This is the callback function called by plugins when
+ * they need a new log line */
+int
+generate_message(char *buffer, int buffer_size, int thread_id, unsigned long seq)
 {
-  res->tv_sec = (t1->tv_sec + t2->tv_sec);
-  res->tv_usec = (t1->tv_usec + t2->tv_usec);
-  if (res->tv_usec >= 1e6)
-    {
-      res->tv_sec++;
-      res->tv_usec -= USEC_PER_SEC;
-    }
-}
+  int str_len;
 
-static ssize_t
-write_chunk(send_data_t send_func, void *send_func_ud, void *buf, size_t buf_len)
-{
-  ssize_t rc;
-  size_t pos = 0;
+  if (read_from_file)
+    str_len = read_next_message_from_file(buffer, buffer_size, syslog_proto, thread_id);
+  else
+    str_len = generate_log_line(buffer, buffer_size, syslog_proto, thread_id, seq);
 
-  while (pos < buf_len)
-    {
-      rc = send_func(send_func_ud, buf + pos, buf_len - pos);
+  g_mutex_lock(message_counter_lock);
```

Comment:
I suggest atomic counters (g_atomic_pointer_add ()) instead of locking.

Pro: this is a heavy path, all threads constantly generate messages. This might give a performance improvement.
Contra: with the lock, the number of messages and the sent messages were corresponding. Atomic counters can include small deviation.

Reply:
Based on latest perf test results it seems the new loggen performs on almost the same level as old version. This means there is no urgent need to speed it up. Otherwise it seems to be very sexy to use atomic counters. I suggest to do this in a next turn. Are you ok with this?